// aws-cdk/lib/api/bootstrap/bootstrap-environment2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cxschema = require("@aws-cdk/cloud-assembly-schema");
const cxapi = require("@aws-cdk/cx-api");
const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const __1 = require("..");
const serialize_1 = require("../../serialize");
const aws_auth_1 = require("../aws-auth");
async function bootstrapEnvironment2(environment, sdkProvider, toolkitStackName, roleArn, props = {}) {
    var _a, _b, _c, _d;
    if (((_a = props.trustedAccounts) === null || _a === void 0 ? void 0 : _a.length) && !((_b = props.cloudFormationExecutionPolicies) === null || _b === void 0 ? void 0 : _b.length)) {
        throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
    }
    const outdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk-bootstrap-new'));
    const builder = new cxapi.CloudAssemblyBuilder(outdir);
    // convert from YAML to JSON (which the Cloud Assembly uses)
    const templateFile = `${toolkitStackName}.template.json`;
    const bootstrapTemplatePath = path.join(__dirname, 'bootstrap-template.yaml');
    const bootstrapTemplateObject = await serialize_1.loadStructuredFile(bootstrapTemplatePath);
    await fs.writeJson(path.join(builder.outdir, templateFile), bootstrapTemplateObject);
    builder.addArtifact(toolkitStackName, {
        type: cxschema.ArtifactType.AWS_CLOUDFORMATION_STACK,
        environment: cxapi.EnvironmentUtils.format(environment.account, environment.region),
        properties: {
            templateFile,
        },
    });
    const resolvedEnvironment = await sdkProvider.resolveEnvironment(environment.account, environment.region);
    const assembly = builder.buildAssembly();
    return await __1.deployStack({
        stack: assembly.getStackByName(toolkitStackName),
        resolvedEnvironment,
        sdk: await sdkProvider.forEnvironment(environment.account, environment.region, aws_auth_1.Mode.ForWriting),
        sdkProvider,
        roleArn,
        tags: props.tags,
        execute: props.execute,
        parameters: {
            FileAssetsBucketName: props.bucketName,
            FileAssetsBucketKmsKeyId: props.kmsKeyId,
            TrustedAccounts: (_c = props.trustedAccounts) === null || _c === void 0 ? void 0 : _c.join(','),
            CloudFormationExecutionPolicies: (_d = props.cloudFormationExecutionPolicies) === null || _d === void 0 ? void 0 : _d.join(','),
        },
    });
}
exports.bootstrapEnvironment2 = bootstrapEnvironment2;
