// aws-cdk/test/api/bootstrap2.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mockDeployStack = jest.fn();
jest.mock('../../lib/api/deploy-stack', () => ({
    deployStack: mockDeployStack,
}));
const bootstrap_environment2_1 = require("../../lib/api/bootstrap/bootstrap-environment2");
const mock_sdk_1 = require("../util/mock-sdk");
describe('Bootstrapping v2', () => {
    const env = {
        account: '123456789012',
        region: 'us-east-1',
        name: 'mock',
    };
    const sdk = new mock_sdk_1.MockSdkProvider();
    test('passes the bucket name as a CFN parameter', async () => {
        await bootstrap_environment2_1.bootstrapEnvironment2(env, sdk, 'mockStack', undefined, {
            bucketName: 'my-bucket-name',
        });
        expect(mockDeployStack).toHaveBeenCalledWith(expect.objectContaining({
            parameters: {
                FileAssetsBucketName: 'my-bucket-name',
            },
        }));
    });
    test('passes the KMS key ID as a CFN parameter', async () => {
        await bootstrap_environment2_1.bootstrapEnvironment2(env, sdk, 'mockStack', undefined, {
            kmsKeyId: 'my-kms-key-id',
        });
        expect(mockDeployStack).toHaveBeenCalledWith(expect.objectContaining({
            parameters: {
                FileAssetsBucketKmsKeyId: 'my-kms-key-id',
            },
        }));
    });
    test('passing trusted accounts without CFN managed policies results in an error', async () => {
        await expect(bootstrap_environment2_1.bootstrapEnvironment2(env, sdk, 'mockStack', undefined, {
            trustedAccounts: ['123456789012'],
        }))
            .rejects
            .toThrow('--cloudformation-execution-policies are required if --trust has been passed!');
    });
    afterEach(() => {
        mockDeployStack.mockClear();
    });
});
