// @aws-cdk/aws-s3/test/test.rules.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'Bucket with expiration days'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            lifecycleRules: [{
                    expiration: core_1.Duration.days(30)
                }]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            LifecycleConfiguration: {
                Rules: [{
                        ExpirationInDays: 30,
                        Status: 'Enabled'
                    }]
            }
        }));
        test.done();
    },
    'Can use addLifecycleRule() to add a lifecycle rule'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const bucket = new lib_1.Bucket(stack, 'Bucket');
        bucket.addLifecycleRule({
            expiration: core_1.Duration.days(30)
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            LifecycleConfiguration: {
                Rules: [{
                        ExpirationInDays: 30,
                        Status: 'Enabled'
                    }]
            }
        }));
        test.done();
    },
    'Bucket with expiration date'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            lifecycleRules: [{
                    expirationDate: new Date('2018-01-01')
                }]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            LifecycleConfiguration: {
                Rules: [{
                        ExpirationDate: '2018-01-01T00:00:00',
                        Status: 'Enabled'
                    }]
            }
        }));
        test.done();
    },
    'Bucket with transition rule'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.Bucket(stack, 'Bucket', {
            lifecycleRules: [{
                    transitions: [{
                            storageClass: lib_1.StorageClass.GLACIER,
                            transitionAfter: core_1.Duration.days(30)
                        }]
                }]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::S3::Bucket', {
            LifecycleConfiguration: {
                Rules: [{
                        Transitions: [{
                                StorageClass: 'GLACIER',
                                TransitionInDays: 30
                            }],
                        Status: 'Enabled'
                    }]
            }
        }));
        test.done();
    },
    'Noncurrent rule on nonversioned bucket fails'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN: Fail because of lack of versioning
        test.throws(() => {
            new lib_1.Bucket(stack, 'Bucket1', {
                lifecycleRules: [{
                        noncurrentVersionExpiration: core_1.Duration.days(10)
                    }]
            });
        });
        // WHEN: Succeeds because versioning is enabled
        new lib_1.Bucket(stack, 'Bucket2', {
            versioned: true,
            lifecycleRules: [{
                    noncurrentVersionExpiration: core_1.Duration.days(10)
                }]
        });
        test.done();
    },
};
