// cdk-assets/test/placeholders.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_assets_schema_1 = require("@aws-cdk/cdk-assets-schema");
const mockfs = require("mock-fs");
const lib_1 = require("../lib");
const mock_aws_1 = require("./mock-aws");
let aws;
beforeEach(() => {
    mockfs({
        '/simple/cdk.out/assets.json': JSON.stringify({
            version: cdk_assets_schema_1.AssetManifestSchema.currentVersion(),
            files: {
                fileAsset: {
                    type: 'file',
                    source: {
                        path: 'some_file'
                    },
                    destinations: {
                        theDestination: {
                            // Absence of region
                            assumeRoleArn: 'arn:aws:role-${AWS::AccountId}',
                            bucketName: 'some_bucket-${AWS::AccountId}-${AWS::Region}',
                            objectKey: 'some_key-${AWS::AccountId}-${AWS::Region}',
                        },
                    },
                },
            },
            dockerImages: {
                dockerAsset: {
                    type: 'docker-image',
                    source: {
                        directory: 'dockerdir'
                    },
                    destinations: {
                        theDestination: {
                            // Explicit region
                            region: 'explicit_region',
                            assumeRoleArn: 'arn:aws:role-${AWS::AccountId}',
                            repositoryName: 'repo-${AWS::AccountId}-${AWS::Region}',
                            imageTag: 'abcdef',
                        },
                    },
                },
            },
        }),
        '/simple/cdk.out/some_file': 'FILE_CONTENTS',
    });
    aws = mock_aws_1.mockAws();
});
afterEach(() => {
    mockfs.restore();
});
test('check that placeholders are replaced', async () => {
    const pub = new lib_1.AssetPublishing(lib_1.AssetManifest.fromPath('/simple/cdk.out'), { aws });
    aws.mockS3.getBucketLocation = mock_aws_1.mockedApiResult({});
    aws.mockS3.headObject = mock_aws_1.mockedApiResult({ /* No error == file exists */});
    aws.mockEcr.describeImages = mock_aws_1.mockedApiResult({ /* No error == image exists */});
    await pub.publish();
    expect(aws.s3Client).toHaveBeenCalledWith(expect.objectContaining({
        assumeRoleArn: 'arn:aws:role-current_account',
    }));
    expect(aws.ecrClient).toHaveBeenCalledWith(expect.objectContaining({
        region: 'explicit_region',
        assumeRoleArn: 'arn:aws:role-current_account',
    }));
    expect(aws.mockS3.headObject).toHaveBeenCalledWith(expect.objectContaining({
        Bucket: 'some_bucket-current_account-current_region',
        Key: 'some_key-current_account-current_region'
    }));
    expect(aws.mockEcr.describeImages).toHaveBeenCalledWith(expect.objectContaining({
        imageIds: [{ imageTag: 'abcdef' }],
        repositoryName: 'repo-current_account-explicit_region'
    }));
});
