// cdk-assets/test/progress.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_assets_schema_1 = require("@aws-cdk/cdk-assets-schema");
const mockfs = require("mock-fs");
const lib_1 = require("../lib");
const mock_aws_1 = require("./mock-aws");
let aws;
beforeEach(() => {
    mockfs({
        '/simple/cdk.out/assets.json': JSON.stringify({
            version: cdk_assets_schema_1.AssetManifestSchema.currentVersion(),
            files: {
                theAsset: {
                    source: {
                        path: 'some_file'
                    },
                    destinations: {
                        theDestination1: {
                            region: 'us-north-50',
                            assumeRoleArn: 'arn:aws:role',
                            bucketName: 'some_bucket',
                            objectKey: 'some_key',
                        },
                        theDestination2: {
                            region: 'us-north-50',
                            assumeRoleArn: 'arn:aws:role',
                            bucketName: 'some_bucket',
                            objectKey: 'some_key2',
                        },
                    },
                },
            },
        }),
        '/simple/cdk.out/some_file': 'FILE_CONTENTS',
    });
    aws = mock_aws_1.mockAws();
    // Accept all S3 uploads as new
    aws.mockS3.getBucketLocation = mock_aws_1.mockedApiResult({});
    aws.mockS3.headObject = mock_aws_1.mockedApiFailure('NotFound', 'File does not exist');
    aws.mockS3.upload = mock_aws_1.mockUpload();
});
afterEach(() => {
    mockfs.restore();
});
test('test listener', async () => {
    const progressListener = new FakeListener();
    const pub = new lib_1.AssetPublishing(lib_1.AssetManifest.fromPath('/simple/cdk.out'), { aws, progressListener });
    await pub.publish();
    const allMessages = progressListener.messages.join('\n');
    // Log mentions asset/destination ids
    expect(allMessages).toContain('theAsset:theDestination1');
    expect(allMessages).toContain('theAsset:theDestination2');
});
test('test abort', async () => {
    const progressListener = new FakeListener(true);
    const pub = new lib_1.AssetPublishing(lib_1.AssetManifest.fromPath('/simple/cdk.out'), { aws, progressListener });
    await pub.publish();
    const allMessages = progressListener.messages.join('\n');
    // We never get to asset 2
    expect(allMessages).not.toContain('theAsset:theDestination2');
});
class FakeListener {
    constructor(doAbort = false) {
        this.doAbort = doAbort;
        this.messages = new Array();
    }
    onPublishEvent(_type, event) {
        this.messages.push(event.message);
        if (this.doAbort) {
            event.abort();
        }
    }
}
