// cdk-assets/test/zipping.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Separate test file since the archiving module doesn't work well with 'mock-fs'
const cdk_assets_schema_1 = require("@aws-cdk/cdk-assets-schema");
const lib_1 = require("../lib");
const bockfs = require("./bockfs");
const mock_aws_1 = require("./mock-aws");
let aws;
beforeEach(() => {
    bockfs({
        '/simple/cdk.out/assets.json': JSON.stringify({
            version: cdk_assets_schema_1.AssetManifestSchema.currentVersion(),
            files: {
                theAsset: {
                    source: {
                        path: 'some_dir',
                        packaging: 'zip'
                    },
                    destinations: {
                        theDestination: {
                            region: 'us-north-50',
                            assumeRoleArn: 'arn:aws:role',
                            bucketName: 'some_bucket',
                            objectKey: 'some_key',
                        },
                    },
                },
            },
        }),
        '/simple/cdk.out/some_dir/some_file': 'FILE_CONTENTS',
    });
    aws = mock_aws_1.mockAws();
    // Accept all S3 uploads as new
    aws.mockS3.headObject = mock_aws_1.mockedApiFailure('NotFound', 'File does not exist');
    aws.mockS3.upload = mock_aws_1.mockUpload();
});
afterEach(() => {
    bockfs.restore();
});
test('Take a zipped upload', async () => {
    const pub = new lib_1.AssetPublishing(lib_1.AssetManifest.fromPath(bockfs.path('/simple/cdk.out')), { aws });
    await pub.publish();
    expect(aws.mockS3.upload).toHaveBeenCalledWith(expect.objectContaining({
        Bucket: 'some_bucket',
        Key: 'some_key',
        ContentType: 'application/zip',
    }));
});
